Adding audience to panes and filtering based on that and user roles

megoth opened the pull request:
This is dependent on solid/solid-ui#112, so do not merge this until it has been merged and released and the PR has been updated with corresponding release version.

megoth committed:
Refactored logic for how to find relevant panes

Making it easier to add extra logic that takes what kind of user the user is

megoth committed:
Adding audience and filtering based on user roles in outlineManager

This is dependent on solid/solid-ui#112, so do not merge this until it has been merged and released and the PR has been updated with corresponding release version.

megoth committed:
Removing scratchPad as planned

Was marked for removal in https://team.inrupt.net/public/mashlib//Panes%20(developer%20and%20power%20user%20mode).pdf

timbl reviewed (changes requested):
It is important that when a user is looking at an object (like a meting or a bank transaction ) where there is a hand-coded view, then they must be given that view, whether or not they are an expert user. It is very bad to deprive them of the proper view and give them a raw data view!   We can I suppose detect that as the highest priority pane in the way the pane system works at the moment,  though we cloudl cahnge that to haveing q values for each pane.

Vinnl commented on defaultPane.js:
```diff
@@ -5,14 +5,17 @@
 
   name: 'default',
 
+  audience: [ns.solid('Developer')],
```
I'm not quite sure what this Pane does, but given that it's named "Default": is this the one that will be visible if no other Pane matches? In other words: will the user get an empty screen if this one's not present either? If so, maybe it should just do what it does today if the user's a Developer, and display a message about why the user doesn't see anything if not.

megoth replied:
I think the name is just misleading - I suspect it was one of the first panes added, and the idea was to give you a little bit of info about the resource. But this info is still designed for developers.

If no panes are found, it will show `internalPane` instead. (This is also noted as developer-oriented, but this is hard-coded to be shown if no other panes are available.)

Vinnl commented on outline/manager.js:
```diff
@@ -490,8 +475,7 @@ module.exports = function (doc) {
 
-                // If the view already exists, remove it
-              state = 'paneShown'
+              // If the view already exists, remove it
```
It's not quite clear to me what line this comment applies to, but if it was the line `state = 'paneShown'`, then I suppose this comment should be removed too :)

Vinnl reviewed (changes requested):
LGTM; setting it to "Request changes" to make sure the upgrade to a new major version of solid-ui with https://github.com/solid/solid-ui/pull/112 is included.

megoth committed:
Removed unnecessary comment

Should've been removed in earlier commit

megoth committed:
Show most relevant pane in some cases

Handle the case that Tim describes in https://github.com/solid/solid-panes/pull/168#pullrequestreview-281536094

timbl commented on outline/manager.js:
```diff
@@ -380,8 +380,21 @@ module.exports = function (doc) {
   async function getRelevantPanes (panes, subject, dom) {
     const relevantPanes = panes.list.filter(pane => pane.label(subject, dom) && !pane.global)
+    if (relevantPanes.length === 0) {
+      // there are no relevant panes, simply return default pane (which ironically is internalPane)
```
¿¿¿ eh??? The `default` pane and the `internals` pane should be quite different.  The default pane should the normal properties in a list, excluding things deemed to be internal like the URI itself, the record of attempts to load the URI etc.  Those things are shown in the internals pane, which also allows you to refresh the local triple store, delete the file, etc.  The default pane is not the internal pane.

megoth replied:
The irony is in the naming - defaultPane sounds like it should be the _default_ pane that is shown.

megoth committed:
Upgraded deps
